pkg/querier/tripperware/instantquery/instant_query.go

```diff
@@ -0,0 +1,442 @@
+	}
+
+	log, ctx := spanlogger.New(ctx, "PrometheusInstantQueryResponse") //nolint:ineffassign,staticcheck
+	defer log.Finish()
+
+	buf, err := tripperware.BodyBuffer(r)
+	if err != nil {
+		log.Error(err)
+		return nil, err
+	}
+	var resp PrometheusInstantQueryResponse
+	if err := json.Unmarshal(buf, &resp); err != nil {
+		return nil, httpgrpc.Errorf(http.StatusInternalServerError, "error decoding response: %v", err)
+	}
+
+	for h, hv := range r.Header {
+		resp.Headers = append(resp.Headers, &tripperware.PrometheusResponseHeader{Name: h, Values: hv})
+	}
+	return &resp, nil
+}
+
+func (instantQueryCodec) EncodeRequest(ctx context.Context, r tripperware.Request) (*http.Request, error) {
+	promReq, ok := r.(*PrometheusRequest)
+	if !ok {
+		return nil, httpgrpc.Errorf(http.StatusBadRequest, "invalid request format")
+	}
+	params := url.Values{
+		"time":  []string{tripperware.EncodeTime(promReq.Time)},
+		"query": []string{promReq.Query},
+	}
+
+	if promReq.Stats != "" {
+		params.Add("stats", promReq.Stats)
+	}
+
+	u := &url.URL{
+		Path:     promReq.Path,
+		RawQuery: params.Encode(),
+	}
+
+	var h = http.Header{}
+
+	for n, hv := range promReq.Headers {
+		for _, v := range hv {
+			h.Add(n, v)
+		}
+	}
+
+	req := &http.Request{
+		Method:     "GET",
+		RequestURI: u.String(), // This is what the httpgrpc code looks at.
+		URL:        u,
+		Body:       http.NoBody,
+		Header:     h,
+	}
+
+	return req.WithContext(ctx), nil
+}
+
+func (instantQueryCodec) EncodeResponse(ctx context.Context, res tripperware.Response) (*http.Response, error) {
+	sp, _ := opentracing.StartSpanFromContext(ctx, "APIResponse.ToHTTPResponse")
+	defer sp.Finish()
+
+	a, ok := res.(*PrometheusInstantQueryResponse)
+	if !ok {
+		return nil, httpgrpc.Errorf(http.StatusInternalServerError, "invalid response format")
+	}
+
+	b, err := json.Marshal(a)
+	if err != nil {
+		return nil, httpgrpc.Errorf(http.StatusInternalServerError, "error encoding response: %v", err)
+	}
+
+	sp.LogFields(otlog.Int("bytes", len(b)))
+
+	resp := http.Response{
+		Header: http.Header{
+			"Content-Type": []string{"application/json"},
+		},
+		Body:          io.NopCloser(bytes.NewBuffer(b)),
+		StatusCode:    http.StatusOK,
+		ContentLength: int64(len(b)),
+	}
+	return &resp, nil
+}
+
+func (instantQueryCodec) MergeResponse(responses ...tripperware.Response) (tripperware.Response, error) {
+	if len(responses) == 0 {
+		return NewEmptyPrometheusInstantQueryResponse(), nil
+	} else if len(responses) == 1 {
+		return responses[0], nil
+	}
+
+	promResponses := make([]*PrometheusInstantQueryResponse, 0, len(responses))
+	for _, resp := range responses {
+		promResponses = append(promResponses, resp.(*PrometheusInstantQueryResponse))
+	}
+
+	var r isPrometheusInstantQueryResult_Result
+	switch promResponses[0].Data.ResultType {
+	case model.ValVector.String():
+		r = &PrometheusInstantQueryResult_Vector{
+			Vector: vectorMerge(promResponses),
+		}
```

Comment:
What about other types? https://prometheus.io/docs/prometheus/latest/querying/api/#instant-queries We also have `matrix`, `scalar` and `string`.

Looks like we don't hanle them since we know `len(responses) <= 1` in this case?

Reply:
As per Thanos implementation, we cannot have any other response other than vector if the query is shardable.

I think i should document that there in a comment.